src/core/strategies/token.js

```diff
@@ -0,0 +1,285 @@
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2018-2023 Open Assessment Technologies SA ;
+ */
+
+import { terms } from '../terms.js';
+import tokensHelper from '../tokens.js';
+
+/**
+ * List of tokens that refuse explicit positive sign
+ * @type {string[]}
+ */
+const refuseExplicitPositive = ['LPAR', 'SUB', 'ADD', 'MUL', 'DIV', 'MOD', 'POW', 'ASSIGN'];
+
+/**
+ * Checks if a token accept an explicit positive sign on the right
+ * @param {token} token
+ * @returns {boolean}
+ */
+const acceptExplicitPositive = token =>
+    !token || (!tokensHelper.isFunction(token) && refuseExplicitPositive.indexOf(token.type) === -1);
+
+/**
+ * Produces a descriptor to insert a negative sign
+ * @param {token} token
+ * @returns {tokenChange}
+ */
+function insertNegativeSign(token) {
+    const { offset } = token;
+    const { value } = terms.SUB;
+    return {
+        offset,
+        length: 0,
+        value,
+        move: value.length
+    };
+}
+
+/**
+ * Produces a descriptor to replace a sign by a negative sign
+ * @param {token} token
+ * @returns {tokenChange}
+ */
+function replaceByNegativeSign(token) {
+    const { offset } = token;
+    const { value } = terms.SUB;
+    return {
+        offset,
+        length: token.value.length,
+        value,
+        move: value.length - token.value.length
+    };
+}
+
+/**
+ * Produces a descriptor to replace a sign by a positive sign
+ * @param {token} token
+ * @param {number} index
+ * @param {token[]} tokens
+ * @returns {tokenChange}
+ */
+function replaceByPositiveSign(token, index, tokens) {
+    const allowExplicit = index && acceptExplicitPositive(tokens[index - 1]);
+    const { offset } = token;
+    const value = (allowExplicit && terms.ADD.value) || '';
+    return {
+        offset,
+        length: token.value.length,
+        value,
+        move: value.length - token.value.length
+    };
+}
+
+/**
+ * Apply a sign change at the current index
+ * @param {number} index - The index of the current token
+ * @param {token[]} tokens - The list of tokens that represent the expression
+ * @returns {tokenChange|null} - The result of the strategy: `null` if cannot apply, or the descriptor of the change
+ */
+function applySignChange(index, tokens) {
+    const token = tokens[index] || null;
+    const nextToken = tokens[index + 1] || null;
+    const type = tokensHelper.getType(token);
```

Comment:
whats the difference betwee tokensHelper.getType(token) and token.type?

Reply:
`tokensHelper.getType(token)` will return the term type, from the list `[digit, operator, variable, function, ...]`

`token.type` is defined by the lexer which is returning a descriptor of this form:
- type: the abstract name of the token, like `LPAR` or `NUM1`
- value: the actual value of the token, like `(` or `1`
- some other info like the position, the actual text like ` (` or ` 1 `, etc.

If you are not familiar with the concept of lexer or formal grammar, I'd advise looking at [Moo.js](https://github.com/no-context/moo), [Nearley.js](https://nearley.js.org/docs/grammar#) or [Jison](http://techtonik.github.io/jison/docs/)